### ISSUE TITLE
Rule proposal: Disallow implicit string conversion in template literals

### ISSUE THREAD

**Issue by ark120202** [typescript-eslint/typescript-eslint#818]
TypeScript allows usage of any types in interpolations, which can lead to unexpected results:

```ts
const func = () => 'text';
`${func}` // => "() => 'text'"

const object = { foo: 'bar' };
`${object}` // => "[object Object]"

let value: string | undefined;
`${value}` // => "undefined"
```

This rule should disallow usage of any types, except primitives that are converted to expected values (strings, numbers and booleans).

There is the same [proposal to tslint](https://github.com/palantir/tslint/issues/4512) that also includes plus operands, but I'm not sure if there's need to include it in this rule, since it appears to be covered by restrict-plus-operands.

**Comment by ColCh** [comment]
Do you want me to help here?

**Comment by bradzacher** [comment]
You can if you'd like. We love all contributors!

This is a pretty advanced rule to implement as your first.
There'll be a lot of pieces you'll have to figure out all at once (this codebase, eslint, and the typescript compiler API).
Feel free to dive right in to it if you'd like, as we don't have any plans to implement this ourselves in the short term.
